### PR TITLE
Dan

### PR DESCRIPTION
**Pull request by Mach131**
based on Extra's branch; unit tests not finished yet

### CODE REVIEW

File: scenes/game/local_game.gd [hunk 1]
```diff
@@ -4132,13 +4188,17 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
 		"draw":
+			var from_bottom = 'from_bottom' in effect and effect['from_bottom']
 			var amount = effect['amount']
 			if str(amount) == "strike_x":
 				amount = performing_player.strike_stat_boosts.strike_x
 			if amount > 0:
 				if 'opponent' in effect and effect['opponent']:
 					events += opposing_player.draw(amount)
 					_append_log_full(Enums.LogType.LogType_CardInfo, opposing_player, "draws %s card(s)." % amount)
+				if from_bottom:
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:4199]: What is going on here?
from_bottom should just be passed in for opponent or self, why is there this extra if?

File: scenes/game/local_game.gd [hunk 2]
```diff
@@ -4381,8 +4447,12 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
 			performing_player.ignore_push_and_pull -= 1
 			if performing_player.ignore_push_and_pull == 0:
 				_append_log_full(Enums.LogType.LogType_Effect, performing_player, "no longer ignores pushes and pulls.")
+		"look_at_top_deck":
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:4450]: can you just make this "reveal_topdeck" to go with reveal_hand and reveal_hand_and_topdeck?
And put it down there of course

File: scenes/game/local_game.gd [hunk 3]
```diff
@@ -4381,8 +4447,12 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
+			events += performing_player.reveal_topdeck()
 		"look_at_top_opponent_deck":
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:4452]: and maybe just got ahead and change this to reveal_topdeck  opponent: true?
there's only 2 effects (testament/anji) to go change it for.

File: scenes/game/local_game.gd [hunk 4]
```diff
@@ -4381,8 +4447,12 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
 			events += opposing_player.reveal_topdeck()
+		"look_at_opponent_hand":
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:4454]: definitely should just be reveal_hand, opponent: true, and go add that down there.

File: scenes/game/local_game.gd [hunk 5]
```diff
@@ -6995,6 +7072,7 @@ func continue_resolve_strike(events):
 			StrikeState.StrikeState_Card1_DetermineHit:
 				var hit = determine_if_attack_hits(events, player1, player2, card1)
 				if hit:
+					active_strike.player1_hit = true
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:7075]: why is this here?
doesn't determine_if_attack_hits do this?

File: scenes/game/local_game.gd [hunk 6]
```diff
@@ -7203,14 +7282,22 @@ func begin_extra_attack(events, performing_player : Player, card_id : int):
...
-	active_strike.extra_attack_in_progress = true
-	active_strike.extra_attack_card = card_db.get_card(card_id)
-	active_strike.extra_attack_player = performing_player
-	active_strike.extra_attack_previous_attack_power_bonus = performing_player.strike_stat_boosts.power
-	active_strike.extra_attack_previous_attack_min_range_bonus = performing_player.strike_stat_boosts.min_range
-	active_strike.extra_attack_previous_attack_max_range_bonus = performing_player.strike_stat_boosts.max_range
-	active_strike.extra_attack_state = ExtraAttackState.ExtraAttackState_PayCosts
-	active_strike.extra_attack_hit = false
+
+	# Also extra attacks can happen inside of extra attacks, so do a linked list thing to handle that
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:7286]: i hope there's a test for this, lol...

File: scenes/game/local_game.gd [hunk 7]
```diff
@@ -7257,61 +7344,67 @@ func continue_extra_attack(events):
...
+				active_strike.extra_attack_data.extra_attack_remaining_effects = get_all_effects_for_timing("before", attacker_player, attacker_card, true, true)
 			ExtraAttackState.ExtraAttackState_Before:
 				events += do_remaining_effects(attacker_player, ExtraAttackState.ExtraAttackState_DetermineHit)
 			ExtraAttackState.ExtraAttackState_DetermineHit:
 				var hit = determine_if_attack_hits(events, attacker_player, defender_player, attacker_card)
 				if hit:
-					active_strike.extra_attack_hit = true
-					active_strike.extra_attack_state = ExtraAttackState.ExtraAttackState_Hit
-					active_strike.extra_attack_remaining_effects = get_all_effects_for_timing("hit", attacker_player, attacker_card, true, true)
+					active_strike.extra_attack_data.extra_attack_hit = true
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:7362]: ok, now i see why you put player1_hit in the regular strike thing.
This makes sense since you don't want to make the previous attack count as hit for going to gauge.

File: scenes/game/local_game.gd (hunk 5 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:7075]: Okay, i saw below where you changed those things. This makes sense.

File: scenes/game/local_game.gd [hunk 8]
```diff
@@ -1705,7 +1727,8 @@ class Player:
 					deck.remove_at(0)
 					events += [parent.create_event(Enums.EventType.EventType_Draw, my_id, card.id)]
 
-			update_public_hand_if_deck_empty()
+			if update_if_empty:
+				update_public_hand_if_deck_empty()
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:1731]: ok, so it's for the choose from topdeck stuff, makes sense.

File: data/decks/dan.json [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+				"and": {
+					"effect_type": "speedup",
+					"amount": 1
+				}
+			}
+		},
+		{
+			"timing": "end_of_turn",
+			"condition": "not_this_turn_was_strike",
+			"effect_type": "skip_end_of_turn_draw",
```
[the remark this answers is not in the document]
Comment (Mach131) [reply data/decks/dan.json:65]: I also realized this would ignore the hand size limit

File: scenes/game/game.gd [hunk 10]
```diff
@@ -2187,6 +2200,22 @@ func _on_reshuffle_discard(event):
 	update_card_counts()
 	return SmallNoticeDelay
 
+func _on_shuffle_deck(event):
```
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/game.gd:2203]: Oh i didn't even see that this was added; I think I deleted the part that sent the event, I'll add it back to do that

File: scenes/game/local_game.gd (hunk 2 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:4450]: actually, i guess there's a slight difference in that reveal goes to both players but look at is just one?
Double check i guess that it makes sense

File: scenes/game/local_game.gd [hunk 11]
```diff
@@ -1693,9 +1713,11 @@ class Player:
 
 		for i in range(num_to_draw):
 			if len(deck) > 0:
-				var card = deck[0]
+				if from_bottom:
```
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/local_game.gd:1716]: Actually apparently deck.remove_at doesn't accept negative indices, so I had to adjust this a bit while leaving it in the loop

File: scenes/game/local_game.gd (hunk 2 above)
Comment (Mach131) [reply scenes/game/local_game.gd:4450]: nah i think they were both literally using the same function

File: scenes/game/local_game.gd [hunk 12]
```diff
@@ -1724,7 +1729,7 @@ class Player:
 				if not parent.game_over:
 					var card = deck[0]
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd]: you didn't change which card was pulled off though?
Maybe in a later commit you fix this, i'm just going through them in order.

Comment (daniel-k-taylor) [reply scenes/game/local_game.gd]: okay, looks like you got it in the latest commit

**Review by daniel-k-taylor — APPROVED** [review]
